### PR TITLE
bpf: nodeport: handle revDNAT for local backends at to-netdev/to-overlay

### LINKED ISSUE #21955: NodePort BPF with L7 netpol is broken for N/S when request is sent to node running SVC backend
First issue is https://github.com/cilium/cilium/issues/21954. The second issue is that the L7 proxy will do the TCP handshake on behalf of the pod meaning that all replies to an outside client won't be rev-DNAT-ed (bpf_lxc is doing rev-DNAT for NodePort BPF). To fix this we should move the rev-DNAT to `bpf_host @ eth0`, as it was suggested by https://github.com/cilium/cilium/pull/17504. 

The bonus issue is DSR when an intermediate LB node is hit first. In this case, the IP opt/ext is dropped by the proxy. The fix is to handle the opt / ext in bpf_host (same as rev-DNAT).

### LINKED ISSUE #22659: Reply by local NodePort backend in hostNetwork doesn't get revNATed
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Steps to repro:
1. run 1-replica deployment with hostNetwork enabled:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
spec:
  selector:
    matchLabels:
      app: nginx
  replicas: 1 
  template:
    metadata:
      labels:
        app: nginx
    spec:
      hostNetwork: true
      containers:
      - name: nginx
        image: nginx:1.14.2
```
3. Create NodePort Service
```
apiVersion: v1
kind: Service
metadata:
  labels:
    app: nginx
  name: nginx
spec:
  ports:
  - port: 8088
    protocol: TCP
    targetPort: 80
  selector:
    app: nginx
  type: NodePort
```
5. Get the service's Port
6. Get the Node on which the Pod from 1. runs
7. curl to the Node that has the Pod from 1. running `curl <node_ip>:<nodeport_port>`

In step 7. curl will timeout.

### Cilium Version

Tested with 1.11.7 and 1.12.4

### Kernel Version

5.4.0-42-generic #46-Ubuntu SMP ... x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

v1.25.3

### Sysdump

N/A

### Relevant log output

```shell
IPs:
* 21.0.100.49 <- client
* 21.0.100.111 <- Node with Pod
* 30622 <- Service's Node Port


tcpdump on the client side:
21:54:33.214939 IP 21.0.100.49.49034 > 21.0.100.111.30622: Flags [S], seq 611078560, win 64240, options [mss 1460,sackOK,TS val 2341570527 ecr 0,nop,wscale 7], length 0
21:54:33.215274 IP 21.0.100.111.80 > 21.0.100.49.49034: Flags [S.], seq 5949333, ack 611078561, win 65160, options [mss 1460,sackOK,TS val 2525517164 ecr 2341570527,nop,wscale 7], length 0
21:54:33.215323 IP 21.0.100.49.49034 > 21.0.100.111.80: Flags [R], seq 611078561, win 0, length 0


Dump from `cilium monitor -v` from the node that runs the Pod (different flow than above):

<- network flow 0x68c2c982 , identity unknown->unknown state unknown ifindex ens192 orig-ip 0.0.0.0: 21.0.100.49:54090 -> 21.0.100.111:30622 tcp SYN
CPU 07: MARK 0x68c2c982 FROM 3011 DEBUG: Successfully mapped addr=21.0.100.49 to identity=2
CPU 07: MARK 0x68c2c982 FROM 3011 DEBUG: Conntrack lookup 1/2: src=21.0.100.49:54090 dst=21.0.100.111:30622
CPU 07: MARK 0x68c2c982 FROM 3011 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=4
CPU 07: MARK 0x68c2c982 FROM 3011 DEBUG: CT verdict: New, revnat=0
CPU 07: MARK 0x68c2c982 FROM 3011 DEBUG: Service backend slot lookup: slot=1, dport=30622
CPU 07: MARK 0x68c2c982 FROM 3011 DEBUG: Conntrack create: proxy-port=0 revnat=27 src-identity=0 lb=0.0.0.0
CPU 07: MARK 0x68c2c982 FROM 3011 DEBUG: Conntrack lookup 1/2: src=21.0.100.49:54090 dst=21.0.100.111:80
CPU 07: MARK 0x68c2c982 FROM 3011 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 07: MARK 0x68c2c982 FROM 3011 DEBUG: CT verdict: New, revnat=0
CPU 07: MARK 0x68c2c982 FROM 3011 DEBUG: Conntrack create: proxy-port=0 revnat=27 src-identity=2 lb=21.0.100.111
-> network flow 0x11d9165b , identity unknown->unknown state unknown ifindex 0 orig-ip 0.0.0.0: 21.0.100.111:80 -> 21.0.100.49:54090 tcp SYN, ACK
```

### Anything else?

This issue is not related to #22557. I have reverted #18815 and was able to curl to the service using different Node. Curl to node with the Pod was still failing.

Tested with tunnel enabled and disabled.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #22838: Unable to get ingress working using hostNetwork without KubeProxy
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I'm trying to set up Cilium in EKS with `kubeProxyReplacement: strict` and `hostNetwork: true` for my ingress-nginx deployment but I can't manage to get the ingress traffic to reach the ingress-nginx backend Pods.  Cilium Helm values:

```
localRedirectPolicy: true
kubeProxyReplacement: strict
k8sServiceHost: ${k8s_service_host}
k8sServicePort: ${k8s_service_port}

nodePort:
  enabled: true

hostPort:
  enabled: true

externalIPs:
  enabled: true

```

Ingress-nginx:
```
❯ kgp -n ingress-nginx -o wide
NAME                                                READY   STATUS    RESTARTS   AGE   IP              NODE                                          NOMINATED NODE   READINESS GATES
ingress-nginx-public-controller-75bd44d497-cjql5    1/1     Running   0          18h   10.100.30.26    ip-10-100-30-26.eu-west-1.compute.internal    <none>           <none>
ingress-nginx-public-controller-75bd44d497-ffg5g    1/1     Running   0          18h   10.100.25.186   ip-10-100-25-186.eu-west-1.compute.internal   <none>           <none>
ingress-nginx-public-controller-75bd44d497-nwbnz    1/1     Running   0          18h   10.100.33.21    ip-10-100-33-21.eu-west-1.compute.internal    <none>           <none>

❯ k get svc -n ingress-nginx 
NAME                                         TYPE           CLUSTER-IP       EXTERNAL-IP                                                                     PORT(S)                      AGE
ingress-nginx-public-controller              LoadBalancer   172.20.211.31    aa078b8965ac345d8adc0db725132b45-13ead0a130f655f7.elb.eu-west-1.amazonaws.com   80:31739/TCP,443:31772/TCP   18h
```

Cilium:
```
❯ k get nodes -o wide
NAME                                          STATUS   ROLES    AGE   VERSION               INTERNAL-IP     EXTERNAL-IP   OS-IMAGE         KERNEL-VERSION                 CONTAINER-RUNTIME
ip-10-100-24-63.eu-west-1.compute.internal    Ready    <none>   18h   v1.24.7-eks-fb459a0   10.100.24.63    <none>        Amazon Linux 2   5.4.219-126.411.amzn2.x86_64   containerd://1.6.6
ip-10-100-25-186.eu-west-1.compute.internal   Ready    <none>   19h   v1.24.7-eks-fb459a0   10.100.25.186   <none>        Amazon Linux 2   5.4.219-126.411.amzn2.x86_64   containerd://1.6.6
ip-10-100-30-26.eu-west-1.compute.internal    Ready    <none>   19h   v1.24.7-eks-fb459a0   10.100.30.26    <none>        Amazon Linux 2   5.4.219-126.411.amzn2.x86_64   containerd://1.6.6
ip-10-100-33-21.eu-west-1.compute.internal    Ready    <none>   19h   v1.24.7-eks-fb459a0   10.100.33.21    <none>        Amazon Linux 2   5.4.219-126.411.amzn2.x86_64   containerd://1.6.6

❯ k exec -ti cilium-pbmnz -- cilium service list | grep 31772 -a2
Defaulted container "cilium-agent" out of: cilium-agent, mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init)
39   10.100.33.21:31772     NodePort        1 => 10.100.33.21:443 (active)      
40   10.100.33.21:31772/i   NodePort        1 => 10.100.33.21:443 (active)      
                                            2 => 10.100.30.26:443 (active)      
                                            3 => 10.100.25.186:443 (active)     
41   0.0.0.0:31772          NodePort        1 => 10.100.33.21:443 (active)      
42   0.0.0.0:31772/i        NodePort        1 => 10.100.33.21:443 (active)      
                                            2 => 10.100.30.26:443 (active)      
                                            3 => 10.100.25.186:443 (active)    
```
When I do curl to the external IP of the AWS LB it just hangs but when doing tcpdump on the K8S nodes I can see the TCP SYN  packages arrive but there is no response. When curl against `10.100.33.21:443`, `10.100.30.26:443` and  `10.100.25.186:443` from the K8S nodes there is connectivity to the backend Pods.

I can also mention when I remove  `hostNetwork: true`  from the nginx deployment, ingress works as expected but I need   `hostNetwork: true` for other reasons.

So the question is if there's some misconfiguration from my side or if this scenario is just not working?

### Cilium Version

1.12.2

### Kernel Version

5.4.219-126.411.amzn2.x86_64

### Kubernetes Version

1.24

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by julianwiedmann**
Replies by local service backends currently get their revDNAT processing in
from-container. There's two problems with that:
1. we have cases where the packet doesn't reach this part of the
   from-container program (ie. redirect to host proxy), and
2. we're not handling replies by hostns backends at all.

So add an additional check for RevDNAT in handle_nat_fwd() at to-netdev,
before such an untranslated reply leaves the node. The same is needed for
to-overlay, see
3a83623b5f94 ("bpf: add support for local NodePort via tunnel").

Fixes: https://github.com/cilium/cilium/issues/22659
Fixes: https://github.com/cilium/cilium/issues/22838
Fixes: https://github.com/cilium/cilium/issues/21955

### CODE REVIEW

**Comment by julianwiedmann** [comment]
/test-1.26-net-next

**Comment by julianwiedmann** [comment]
/test-1.26-net-next

**Comment by julianwiedmann** [comment]
This isn't quite ready for merge yet (commit messages + tests need more polish), but the datapath code should be good as-is. So let's get the review started ...

**Comment by julianwiedmann** [comment]
@dylandreimerink forgot the doc update for `PKTGEN` of course :slightly_smiling_face: . Queued it up for the next push ...

**Comment by julianwiedmann** [comment]
Some late updates:
- added trace aggregation in the revDNAT path
- polished for the commit messags
- added docs for `PKTGEN` in BPF unit tests
- added various XDP unit tests for LB in NAT mode
- temporarily included the fix from https://github.com/cilium/cilium/pull/23119, to unblock the tests.

**Review by aanm — APPROVED** [review]

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -2180,20 +2354,26 @@ declare_tailcall_if(__or3(__and(is_defined(ENABLE_IPV4),
...
 int tail_handle_nat_fwd_ipv4(struct __ctx_buff *ctx)
 {
+	struct trace_ctx trace = {
+		.reason = TRACE_REASON_UNKNOWN,
+		.monitor = TRACE_PAYLOAD_LEN,
+	};
 	int ret;
 	enum trace_point obs_point;
 
+
```
Comment (qmonnet) [inline bpf/lib/nodeport.h]: Checkpatch complains about this blank line, please remove.

File: test/bpf_tests/bpf_test.go [hunk 2]
```diff
@@ -351,6 +355,22 @@ func subTest(progSet programSet, resultMap *ebpf.Map) func(t *testing.T) {
 		// create ctx with the max allowed size(4k - head room - tailroom)
 		ctx := make([]byte, 4096-256-320)
 
+		if progSet.pktgenProg != nil {
+			statusCode, result, err := progSet.pktgenProg.Test(ctx)
+			if err != nil {
+				t.Fatalf("error while running setup prog: %s", err)
```
Comment (qmonnet) [inline test/bpf_tests/bpf_test.go]:
```suggestion
				t.Fatalf("error while running pktgen prog: %s", err)
```

File: bpf/tests/tc_nodeport_lb4_nat_lb.c [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+		test_fatal("src port has changed");
+
+	if (l4->dest != BACKEND_PORT)
+		test_fatal("dst TCP port hasn't been NATed to backend port");
+
+	test_finish();
+}
+
+/* Test that a reply by the local backend gets revDNATed at to-netdev
+ */
```
Comment (qmonnet) [inline bpf/tests/tc_nodeport_lb4_nat_lb.c]:
```suggestion
/* Test that a reply by the local backend gets revDNATed at to-netdev */
```

File: bpf/tests/tc_nodeport_lb4_nat_backend.c [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+		.scope = LB_LOOKUP_SCOPE_EXT,
+	};
+	/* Create a service with only one backend */
+	struct lb4_service lb_svc_value = {
+		.count = 1,
+		.flags = SVC_FLAG_ROUTABLE,
+	};
+	map_update_elem(&LB4_SERVICES_MAP_V2, &lb_svc_key, &lb_svc_value, BPF_ANY);
+	/* We need to register both in the external and internal scopes for the */
+	/* packet to be redirected to a neighboring node */
```
Comment (qmonnet) [inline bpf/tests/tc_nodeport_lb4_nat_backend.c]: Nit: Netdev-style comment markers? Here and elsewhere

**Review by qmonnet — CHANGES_REQUESTED** [review]
A few minor nits, but looks good otherwise. Thanks!

**Review by dylandreimerink — APPROVED** [review]
Nice work on the tests, looking good!

**Review by aspsk — APPROVED** [review]
Looks great (I didn't look at the tests)

File: bpf/tests/tc_nodeport_lb4_nat_backend.c (hunk 4 above)
Comment (julianwiedmann) [reply bpf/tests/tc_nodeport_lb4_nat_backend.c]: Heh - yes, there's lots of copy & paste boilerplate in here. Should be cleaned up now :+1:.

**Comment by julianwiedmann** [comment]
> A few minor nits, but looks good otherwise. Thanks!

@qmonnet thanks, picked all of them up!

**Comment by julianwiedmann** [comment]
Rebase to pick up https://github.com/cilium/cilium/pull/23119, now that it's merged.

**Review by qmonnet — APPROVED** [review]
Looks good, nice work!

**Comment by julianwiedmann** [comment]
test-1.16-4.9 [flaked](https://github.com/cilium/cilium/issues/15455#issuecomment-1387137283) in `K8sDatapathConfig Host firewall With VXLAN` with

> error: unable to upgrade connection: Authorization error (user=kube-apiserver-kubelet-client, verb=create, resource=nodes, subresource=proxy)

and apparently took down the following tests. Rerunning.

**Comment by julianwiedmann** [comment]
/test-1.16-4.9

**Comment by julianwiedmann** [comment]
Reviews are in, CI is green except for the [bpf-next fails](https://github.com/cilium/cilium/actions/runs/3947520228) in `ci-datapath` that https://github.com/cilium/cilium/pull/23171 tracks.

> ❌ 5/31 tests failed (12/237 actions), 0 tests skipped, 0 scenarios skipped:
> connectivity test failed: 5 tests failed
> Test [no-policies]:
>   ❌ no-policies/pod-to-world/http-to-one.one.one.one-0: cilium-test/client-755fb678bd-9fswv (10.244.1.221) -> one.one.one.one-http (one.one.one.one:80)
>   ❌ no-policies/pod-to-world/https-to-one.one.one.one-0: cilium-test/client-755fb678bd-9fswv (10.244.1.221) -> one.one.one.one-https (one.one.one.one:443)
>   ❌ no-policies/pod-to-world/https-to-one.one.one.one-index-0: cilium-test/client-755fb678bd-9fswv (10.244.1.221) -> one.one.one.one-https-index (one.one.one.one:443)
>   ❌ no-policies/pod-to-world/http-to-one.one.one.one-1: cilium-test/client2-5b97d7bc66-4qkrg (10.244.1.9) -> one.one.one.one-http (one.one.one.one:80)
>   ❌ no-policies/pod-to-world/https-to-one.one.one.one-1: cilium-test/client2-5b97d7bc66-4qkrg (10.244.1.9) -> one.one.one.one-https (one.one.one.one:443)
>   ❌ no-policies/pod-to-world/https-to-one.one.one.one-index-1: cilium-test/client2-5b97d7bc66-4qkrg (10.244.1.9) -> one.one.one.one-https-index (one.one.one.one:443)
> Test [to-entities-world]:
>   ❌ to-entities-world/pod-to-world/http-to-one.one.one.one-0: cilium-test/client-755fb678bd-9fswv (10.244.1.221) -> one.one.one.one-http (one.one.one.one:80)
>   ❌ to-entities-world/pod-to-world/http-to-one.one.one.one-1: cilium-test/client2-5b97d7bc66-4qkrg (10.244.1.9) -> one.one.one.one-http (one.one.one.one:80)
> Test [client-egress-l7]:
>   ❌ client-egress-l7/pod-to-world/http-to-one.one.one.one-0: cilium-test/client2-5b97d7bc66-4qkrg (10.244.1.9) -> one.one.one.one-http (one.one.one.one:80)
> Test [client-egress-l7-named-port]:
>   ❌ client-egress-l7-named-port/pod-to-world/http-to-one.one.one.one-1: cilium-test/client2-5b97d7bc66-4qkrg (10.244.1.9) -> one.one.one.one-http (one.one.one.one:80)
> Test [to-fqdns]:
>   ❌ to-fqdns/pod-to-world/http-to-one.one.one.one-0: cilium-test/client-755fb678bd-9fswv (10.244.1.221) -> one.one.one.one-http (one.one.one.one:80)
>   ❌ to-fqdns/pod-to-world/http-to-one.one.one.one-1: cilium-test/client2-5b97d7bc66-4qkrg (10.244.1.9) -> one.one.one.one-http (one.one.one.one:80)
> Error: Process completed with exit code 1.

**Comment by liuyuan10** [comment]
I see it's going to be backported to v1.12 and v1.13. Are there plan for v1.11?

**Comment by julianwiedmann** [comment]
> I see it's going to be backported to v1.12 and v1.13. Are there plan for v1.11?

No objection from my side :). Let's see how smooth the `v1.12` backport goes...

I take it you already tested  on mainline that this fixes your issue?

**Comment by liuyuan10** [comment]
We just made the decision to update to 1.12 so we don't need this fix in v1.11. I haven't tested it but we will see.

@aanm Regarding v1.12 backport, is there ETA for that. ( I see 1.13 is already done).

**Comment by aanm** [comment]
@liuyuan10 No ETA for it.

**Pull request merged**